daemon/src/lib.rs

```diff
@@ -0,0 +1,655 @@
+///   dropped before the process is about to exit.
+///
+/// You can use this with `clap` to parse command-line arguments and daemonize based on them, e.g.:
+///
+/// ```
+/// use clap::Parser;
+/// use daemon::Daemonize;
+///
+/// let (daemonizing, _cleanup) = unsafe { Daemonize::parse().start() };
+/// // ...anything else needed before the parent exits...
+/// daemonizing.finish();
+/// ```
+///
+/// If the process doesn't have anything to do before completing daemonization, then it can call
+/// `run` to start and immediately finish daemonizing.  It still needs to keep around the `Cleanup`
+/// object:
+///
+/// ```
+/// use clap::Parser;
+/// use daemon::Daemonize;
+///
+/// let _cleanup = unsafe { Daemonize::parse().run() };
+/// ```
+///
+/// If the daemon needs to parse additional options beyond those needed just for daemonization,
+/// you can flatten `Daemonize` into a larger `Args` structure, e.g.:
+///
+/// ```
+/// use clap::Parser;
+/// use daemon::Daemonize;
+///
+/// #[derive(Parser, Debug)]
+/// #[clap(version, about)]
+/// struct Args {
+///     #[clap(flatten)]
+///     daemonize: Daemonize,
+///
+///     //...other options...
+/// }
+///
+/// let Args { daemonize, .. } = Args::parse();
+/// let (daemonizing, _cleanup) = unsafe { daemonize.start() };
+/// // ...anything else needed before the parent exits...
+/// daemonizing.finish();
+/// ```
+///
+
+/// # Safety
+///
+/// It's important to call [`Daemonize::start()`] or [`Daemonize::run()`] as soon as possible at
+/// the beginning of a program.  Forking is unsafe because any threads other than the one that
+/// calls `fork` will be dead in the child, and `fork` has other undesirable consequences, too.
+/// Daemonizing will assert-fail if any additional threads have been started.
+///
+/// About all a program should do beforehand is to parse command-line options (needed to determine
+/// whether to detach) and initialize logging (because daemonization will emit log entries).
+#[derive(Clone, Debug, Default, Parser, PartialEq, Eq)]
+pub struct Daemonize {
+    /// Detach from foreground session
+    #[clap(long)]
+    pub detach: bool,
+
+    /// Create monitoring process
+    #[clap(long)]
+    pub monitor: bool,
+
+    /// Do not change directory to root
+    #[clap(long)]
+    pub no_chdir: bool,
+
+    /// Create pidfile
+    #[clap(long)]
+    pub pidfile: Option<PathBuf>,
+}
+
+impl Daemonize {
+    pub fn new() -> Daemonize {
+        Self::default()
+    }
+
+    /// Daemonizes the current process based on the configured options.  See the module comment for
+    /// full usage.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it forks and exits in the parent: any threads, other than
+    /// the calling thread, will be dead in the child.  Thus, it is only safe to call this function
+    /// while the program is single-threaded, and this function will assert-fail if additional
+    /// threads have been started.
```

Comment:
so I guess you are checking some of these invariants.

Reply:
Yeah. There are other weird special cases, like the child doesn't inherit file locks.